c2rust-transpile/src/translator/variadic.rs

```diff
@@ -9,130 +9,110 @@ pub enum VaPart {
+        match_vastart_struct()
+            .or_else(match_vastart_pointer)
+            .or_else(match_vastart_struct_member)
+            .or_else(match_vastart_struct_pointer_member)
     }
 
     pub fn match_vaend(&self, expr: CExprId) -> Option<CDeclId> {
         self.match_vastart(expr)
     }
 
     pub fn match_vacopy(&self, dst_expr: CExprId, src_expr: CExprId) -> Option<(CDeclId, CDeclId)> {
-        let dst_id = self.match_vastart(dst_expr);
-        let src_id = self.match_vastart(src_expr);
-        if let (Some(did), Some(sid)) = (dst_id, src_id) {
-            return Some((did, sid));
-        }
-        None
+        let dst_id = self.match_vastart(dst_expr)?;
+        let src_id = self.match_vastart(src_expr)?;
+        Some((dst_id, src_id))
     }
 
     pub fn match_vapart(&self, expr: CExprId) -> Option<VaPart> {
-        match_or! { [self.ast_context[expr].kind]
-        CExprKind::Call(_, func, ref args) => (func, args) }
-        match_or! { [self.ast_context[func].kind]
-        CExprKind::ImplicitCast(_, fexp, CastKind::BuiltinFnToFnPtr, _, _) => fexp }
-        match_or! { [self.ast_context[fexp].kind]
-        CExprKind::DeclRef(_, decl_id, _) => decl_id }
-        match_or! { [self.ast_context[decl_id].kind]
-        CDeclKind::Function { ref name, .. } => name }
-        match name as &str {
-            "__builtin_va_start" => {
-                if args.len() != 2 {
-                    return None;
-                }
-                self.match_vastart(args[0]).map(VaPart::Start)
-            }
-
-            "__builtin_va_copy" => {
-                if args.len() != 2 {
-                    return None;
-                }
-                self.match_vacopy(args[0], args[1])
-                    .map(|(did, sid)| VaPart::Copy(did, sid))
-            }
-
-            "__builtin_va_end" => {
-                if args.len() != 1 {
-                    return None;
-                }
-                self.match_vaend(args[0]).map(VaPart::End)
-            }
-
+        let (func, args) = match_or! { [self.ast_context[expr].kind]
+        CExprKind::Call(_, func, ref args) => (func, args) }?;
+        let fexp = match_or! { [self.ast_context[func].kind]
+        CExprKind::ImplicitCast(_, fexp, CastKind::BuiltinFnToFnPtr, _, _) => fexp }?;
+        let decl_id = match_or! { [self.ast_context[fexp].kind]
+        CExprKind::DeclRef(_, decl_id, _) => decl_id }?;
+        let name = match_or! { [self.ast_context[decl_id].kind]
+        CDeclKind::Function { ref name, .. } => name }?;
+        let name = name.strip_prefix("__builtin_va_")?;
+        match (name, args.as_slice()) {
+            ("start", &[expr, _]) => self.match_vastart(expr).map(VaPart::Start),
+            ("copy", &[dst_expr, src_expr]) => self
+                .match_vacopy(dst_expr, src_expr)
+                .map(|(did, sid)| VaPart::Copy(did, sid)),
+            ("end", &[expr]) => self.match_vaend(expr).map(VaPart::End),
```

Comment:
NAK on factoring out the `"__builtin_va_"` prefix, because these are magic strings that are good to be able to grep.

Reply:
What do you mean by NAK?

I agree with your point, though.  I can remove this commit.

Reply:
Fixed in https://github.com/immunant/c2rust/pull/638/commits/b4428a365092f531b1c034b69e0d038ad404a355.  Explanation in the revert commit.